samza-core/src/main/scala/org/apache/samza/system/SystemConsumers.scala

```diff
@@ -258,7 +259,14 @@ class SystemConsumers (
 
     trace("Getting fetch map for system: %s" format systemName)
 
-    val systemFetchSet = emptySystemStreamPartitionsBySystem.get(systemName)
```

Comment:
Can we keep `emptySSPBySystem` and `endOfStreamSSPs` separate concepts? (instead of removing an end of stream SSP from the set of empty SSPs)

Atleast the semantic meaning of emptySSPs == the set of SSPs that we have no buffered messages for, and endOfStreamSSPs == the set of SSPs that are at EOS is maintained.

If the only intention is to prevent that SSP from being polled we can add a check during poll to filter out end of stream SSPs. In other words, an SSP is polled iff. both these conditions are true:

If we don't have buffered messages for it
If we have not reached end of stream